cachix/src/Cachix/Client/PushQueue.hs

```diff
@@ -0,0 +1,73 @@
+{- Implements a queue with the following properties:
+
+- waits for queue to be fully pushed when exiting using ctrl-c (SIGINT)
+- allows stopping the producer
+- avoid duplicate pushing of the same store paths
+
+To safetly exit on demand, signal SIGINT.
+-}
+module Cachix.Client.PushQueue
+  ( pushQueue,
+    Queue,
+  )
+where
+
+import qualified Cachix.Client.Push as Push
+import Control.Concurrent.Async
+import Control.Concurrent.STM (TVar, modifyTVar', newTVarIO, readTVarIO)
+import qualified Control.Concurrent.STM.TBQueue as TBQueue
+import qualified Data.Set as S
+import Protolude
+import qualified System.Posix.Signals as Signals
+
+type StorePath = Text
+
+type Queue = TBQueue.TBQueue StorePath
+
+data WorkerState
+  = WorkerState
+      { inQueue :: Queue,
+        inProgress :: TVar (S.Set StorePath)
+      }
+
+worker :: Push.PushCache IO () -> WorkerState -> IO ()
+worker pushCache workerState = forever $ do
+  storePath <- atomically $ TBQueue.readTBQueue $ inQueue workerState
+  inprogress <- readTVarIO $ inProgress workerState
+  -- TODO: there's a small window between checking membership and inserting new path
+  unless (S.member storePath inprogress) $ bracket_ (start storePath) (stop storePath) $
+    Push.pushSingleStorePath pushCache storePath
+  where
+    start storePath =
+      atomically $ modifyTVar' (inProgress workerState) (S.insert storePath)
+    stop storePath =
+      atomically $ modifyTVar' (inProgress workerState) (S.delete storePath)
+
+-- NOTE: producer is responsible for signaling SIGINT upon termination
+pushQueue :: Int -> (Queue -> IO (IO ())) -> Push.PushCache IO () -> IO ()
+pushQueue numWorkers mkProducer pushCache = do
+  queue <- atomically $ TBQueue.newTBQueue 10000
+  stopProducerCallback <- mkProducer queue
+  progress <- newTVarIO S.empty
+  let workerState = WorkerState queue progress
+  consumer <- async $ replicateConcurrently_ numWorkers $ worker pushCache workerState
+  void $ Signals.installHandler Signals.sigINT (Signals.CatchOnce (exitOnceQueueIsEmpty stopProducerCallback consumer workerState)) Nothing
+  wait consumer
+
+exitOnceQueueIsEmpty :: IO () -> Async () -> WorkerState -> IO ()
+exitOnceQueueIsEmpty stopProducerCallback consumer workerState = do
+  putText "Stopping producer of store paths and waiting for queue to empty ..."
+  stopProducerCallback
+  go
+  where
+    go = do
+      queueLength <- atomically $ TBQueue.lengthTBQueue $ inQueue workerState
+      inprogress <- readTVarIO $ inProgress workerState
+      if queueLength == 0 && S.null inprogress
+        then do
+          putText "All done, exiting."
+          uninterruptibleCancel consumer
```

Comment:
Potential race condition here. It's nicer to put an "EOF" on the queue, so the threads can exit by themselves and you just wait for the consumer async to complete here.

Reply:
What is the race condition that can happen?

Reply:
I guess this is more of a general advice than specific to cachix.

 - thread a creates thread b
 - thread b adds to queue
 - thread a terminates the worker
 - thread a assumes everything is done
 - thread b adds to queue

It probably doesn't apply here.

Reply:
On line 60 it terminates the producer, I should note that it should be a blocking action.

I tried the approach with EOF but it makes it more complicated to report number of items in queue.